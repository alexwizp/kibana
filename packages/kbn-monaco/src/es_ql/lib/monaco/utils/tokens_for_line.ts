/*
 * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
 * or more contributor license agreements. Licensed under the Elastic License
 * 2.0 and the Server Side Public License, v 1; you may not use this file except
 * in compliance with, at your election, the Elastic License 2.0 or the Server
 * Side Public License, v 1.
 */

import { CharStreams } from 'antlr4ts';
import { monaco } from '../../../../monaco_imports';

import { ESQLErrorListener } from '../../error_listener';
import { es_ql_lexer as ESQLLexer } from '../../../antlr/es_ql_lexer';
import { ESQLToken } from '../esql_token';
import { ESQLLineTokens } from '../esql_line_tokens';
import { tokenPostfix } from '../esql_constants';

const EOF = -1;

export function tokensForLine(input: string): monaco.languages.ILineTokens {
  const errorStartingPoints: number[] = [];
  const inputStream = CharStreams.fromString(input);
  const lexer = new ESQLLexer(inputStream);

  lexer.removeErrorListeners();
  const errorListener = new ESQLErrorListener();
  lexer.addErrorListener(errorListener);

  let done = false;
  const myTokens: monaco.languages.IToken[] = [];

  do {
    const token = lexer.nextToken();
    if (token == null) {
      done = true;
    } else {
      // We exclude EOF
      if (token.type === EOF) {
        done = true;
      } else {
        const tokenTypeName = lexer.ruleNames[token.type];
        const myToken = new ESQLToken(tokenTypeName, token.startIndex);
        myTokens.push(myToken);
      }
    }
  } while (!done);

  for (const e of errorStartingPoints) {
    myTokens.push(new ESQLToken('error' + tokenPostfix, e));
  }

  myTokens.sort((a, b) => (a.startIndex > b.startIndex ? 1 : -1));

  return new ESQLLineTokens(myTokens);
}
