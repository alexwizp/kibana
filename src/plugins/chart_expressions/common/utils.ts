/*
 * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
 * or more contributor license agreements. Licensed under the Elastic License
 * 2.0 and the Server Side Public License, v 1; you may not use this file except
 * in compliance with, at your election, the Elastic License 2.0 or the Server
 * Side Public License, v 1.
 */

import type { KibanaExecutionContext } from '@kbn/core-execution-context-common';
import type { ExpressionRenderContext } from './types';

export const extractRenderContext = (
  context?: KibanaExecutionContext
): ExpressionRenderContext | undefined => {
  if (context) {
    const recursiveGet = (item: KibanaExecutionContext): KibanaExecutionContext | undefined => {
      if (item.child) {
        return recursiveGet(item.child);
      } else {
        return item.child;
      }
    };

    const deepestChild = recursiveGet(context);

    if (deepestChild) {
      return {
        originatingApp: deepestChild.type,
      };
    }
  }
};
