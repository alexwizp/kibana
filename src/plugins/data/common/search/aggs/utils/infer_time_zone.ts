/*
 * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
 * or more contributor license agreements. Licensed under the Elastic License
 * 2.0 and the Server Side Public License, v 1; you may not use this file except
 * in compliance with, at your election, the Elastic License 2.0 or the Server
 * Side Public License, v 1.
 */

import moment from 'moment';
import { AggTypesDependencies, DataViewField, DataView } from '../../..';

const DEFAULT_TIME_ZONE = 'UTC';

export const getConfigTimeZone = (
  getExecutionContext: AggTypesDependencies['getExecutionContext'],
  getConfig: AggTypesDependencies['getConfig']
) => {
  const configTimezone = getConfig<string>('dateFormat:tz');
  const isDefaultTimezone = configTimezone === 'Browser';

  if (isDefaultTimezone) {
    const { performedOn } = getExecutionContext();

    if (performedOn === 'server') {
      return DEFAULT_TIME_ZONE;
    }

    // If the typeMeta data index template does not have a timezone assigned to the selected field, use the configured tz
    const detectedTimezone = moment.tz.guess();
    const tzOffset = moment().format('Z');

    return detectedTimezone || tzOffset;
  }

  return configTimezone;
};

export function inferTimeZone(
  params: { field?: DataViewField | string; time_zone?: string },
  dataView: DataView,
  aggName: 'date_histogram' | 'date_range',
  getExecutionContext: AggTypesDependencies['getExecutionContext'],
  getConfig: AggTypesDependencies['getConfig']
) {
  let tz = params.time_zone;

  if (!tz && params.field) {
    // If a field has been configured check the index pattern's typeMeta if a date_histogram on that
    // field requires a specific time_zone
    const fieldName = typeof params.field === 'string' ? params.field : params.field.name;

    tz = dataView.typeMeta?.aggs?.[aggName]?.[fieldName]?.time_zone;
  }

  if (!tz) {
    const configTimezone = getConfig<string>('dateFormat:tz');
    const isDefaultTimezone = configTimezone === 'Browser';

    if (isDefaultTimezone) {
      return getConfigTimeZone(getExecutionContext, getConfig);
    }
  }

  return tz || DEFAULT_TIME_ZONE;
}
