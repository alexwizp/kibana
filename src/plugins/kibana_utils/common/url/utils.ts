/*
 * Licensed to Elasticsearch B.V. under one or more contributor
 * license agreements. See the NOTICE file distributed with
 * this work for additional information regarding copyright
 * ownership. Elasticsearch B.V. licenses this file to you under
 * the Apache License, Version 2.0 (the "License"); you may
 * not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *    http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing,
 * software distributed under the License is distributed on an
 * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
 * KIND, either express or implied.  See the License for the
 * specific language governing permissions and limitations
 * under the License.
 */
import { transform, isUndefined } from 'lodash';
import { stringify, parse, StringifyOptions } from 'query-string';
import { encodeQueryComponent } from './encode_query_component';

export const parseUrlQuery = <TReturn = Record<string, any>>(val: string): TReturn => {
  return (parse(val, { sort: false }) as unknown) as TReturn;
};

export const stringifyUrlQuery = (
  query: Record<string, any>,
  encodeFunction:
    | false
    | ((val: string, pctEncodeSpaces?: boolean) => string) = encodeQueryComponent,
  options: StringifyOptions = {}
) => {
  const encodedQuery =
    encodeFunction &&
    transform(query, (result, value, key) => {
      if (key) {
        result[key] = encodeFunction(isUndefined(value) ? '' : value, true);
      }
    });

  return stringify(encodedQuery || query, {
    strict: false,
    encode: false,
    sort: false,
    ...options,
  });
};
