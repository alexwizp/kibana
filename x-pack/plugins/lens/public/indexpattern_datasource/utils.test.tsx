/*
 * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
 * or more contributor license agreements. Licensed under the Elastic License
 * 2.0; you may not use this file except in compliance with the Elastic License
 * 2.0.
 */

import { getPrecisionErrorWarningMessages } from './utils';
import type { IndexPatternPrivateState } from './types';
import type { FramePublicAPI } from '../types';

describe('indexpattern_datasource utils', () => {
  describe('getPrecisionErrorWarningMessages', () => {
    let state: IndexPatternPrivateState;
    let framePublicAPI: FramePublicAPI;

    beforeEach(() => {
      state = {} as IndexPatternPrivateState;
      framePublicAPI = {
        activeData: {
          id: {
            columns: [
              {
                meta: {
                  sourceParams: {
                    hasPrecisionError: false,
                  },
                },
              },
            ],
          },
        },
      } as unknown as FramePublicAPI;
    });
    test('should not show precisionError if hasPrecisionError is false', () => {
      expect(getPrecisionErrorWarningMessages(state, framePublicAPI)).toMatchInlineSnapshot(
        `Array []`
      );
    });

    test('should not show precisionError if hasPrecisionError is not defined', () => {
      delete framePublicAPI.activeData!.id.columns[0].meta.sourceParams!.hasPrecisionError;

      expect(getPrecisionErrorWarningMessages(state, framePublicAPI)).toMatchInlineSnapshot(
        `Array []`
      );
    });

    test('should show precisionError if hasPrecisionError is true', () => {
      framePublicAPI.activeData!.id.columns[0].meta.sourceParams!.hasPrecisionError = true;

      expect(getPrecisionErrorWarningMessages(state, framePublicAPI)).toMatchInlineSnapshot(`
        Array [
          <FormattedMessage
            defaultMessage="{docCount} values for a terms aggregation may be approximate. As a result, any sub-aggregations on the terms aggregation may also be approximate."
            id="xpack.lens.indexPattern.precisionErrorWarning"
            values={
              Object {
                "docCount": <strong>
                  <FormattedMessage
                    defaultMessage="doc_count"
                    id="xpack.lens.indexPattern.precisionErrorWarning.docCount"
                    values={Object {}}
                  />
                </strong>,
              }
            }
          />,
        ]
      `);
    });
  });
});
