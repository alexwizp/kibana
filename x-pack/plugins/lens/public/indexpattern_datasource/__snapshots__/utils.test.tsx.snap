// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`indexpattern_datasource utils cloneLayer should replace ids 1`] = `
Object {
  "a": Object {
    "columnOrder": Array [
      "899ee4b6-3147-4d45-94bf-ea9c02e55d28",
      "ae62cfc8-faa5-4096-a30c-f92ac59922a0",
    ],
    "columns": Object {
      "899ee4b6-3147-4d45-94bf-ea9c02e55d28": Object {
        "params": Object {
          "orderBy": Object {
            "columnId": "ae62cfc8-faa5-4096-a30c-f92ac59922a0",
            "type": "column",
          },
          "orderDirection": "desc",
        },
      },
      "ae62cfc8-faa5-4096-a30c-f92ac59922a0": Object {
        "params": Object {
          "emptyAsNull": true,
          "justForTest": Array [
            "899ee4b6-3147-4d45-94bf-ea9c02e55d28",
            "ae62cfc8-faa5-4096-a30c-f92ac59922a0",
          ],
          "justForTest2": Array [
            Object {
              "899ee4b6-3147-4d45-94bf-ea9c02e55d28": Array [
                Object {
                  "c": "ae62cfc8-faa5-4096-a30c-f92ac59922a0",
                },
              ],
              "a": "899ee4b6-3147-4d45-94bf-ea9c02e55d28",
            },
          ],
        },
      },
    },
    "incompleteColumns": Object {},
    "indexPatternId": "ff959d40-b880-11e8-a6d9-e546fe2bba5f",
  },
  "b": Object {
    "columnOrder": Array [
      "899ee4b6-3147-4d45-94bf-ea9c02e55d28C",
      "ae62cfc8-faa5-4096-a30c-f92ac59922a0C",
    ],
    "columns": Object {
      "899ee4b6-3147-4d45-94bf-ea9c02e55d28C": Object {
        "params": Object {
          "orderBy": Object {
            "columnId": "ae62cfc8-faa5-4096-a30c-f92ac59922a0C",
            "type": "column",
          },
          "orderDirection": "desc",
        },
      },
      "ae62cfc8-faa5-4096-a30c-f92ac59922a0C": Object {
        "params": Object {
          "emptyAsNull": true,
          "justForTest": Array [
            "899ee4b6-3147-4d45-94bf-ea9c02e55d28C",
            "ae62cfc8-faa5-4096-a30c-f92ac59922a0C",
          ],
          "justForTest2": Array [
            Object {
              "899ee4b6-3147-4d45-94bf-ea9c02e55d28C": Array [
                Object {
                  "c": "ae62cfc8-faa5-4096-a30c-f92ac59922a0C",
                },
              ],
              "a": "899ee4b6-3147-4d45-94bf-ea9c02e55d28C",
            },
          ],
        },
      },
    },
    "incompleteColumns": Object {},
    "indexPatternId": "ff959d40-b880-11e8-a6d9-e546fe2bba5f",
  },
}
`;
